### DISCUSSION TITLE
Workplace for Occurrence Rates Project

### DISCUSSION THREAD

**Discussion by catrionamurray** [catrionamurray/occultence#4] [General]
**Notes from ExoV Meeting (20th June 2024)**

@florian-lienhard has Zooniverse stats for original spec sample of ~150 targets
The plan is for 2 coordinated papers

**Timeline**
15th July - @florian-lienhard  finishes RV paper
Ideally by next meeting - **26th July 5pm** we want to:
- [x] - Add flare detection (_Catriona_)
- [ ] - Run a small sample to check everything works  (_Catriona + Florian_)
- [x] - Speed up code! (currently it takes 90d to run 180 targets with a few thousand inj-rec) (_Catriona_)
- [ ] - Check number of SPC targets with Elsa
- [ ] - Decide on sample (_Catriona + Florian_)
- [ ] - Getting the data - @florian-lienhard has code to download data from portal already (_Florian/Catriona?_)
- [ ] - Binning & uncertainties? (_Florian?/Catriona_)
- [x] - Run rotation_and_flares.ipynb with real data + fake data. Test flare injection/recovery + lomb scargle. (_Florian_)

**Longer timeline:**
- [ ] - Rotation/Flare repeat + inclusion of bigger sample. How can we make this less manual? (_Catriona_)
- [ ] - More quality checks / outlier checks (_Catriona_)
- [ ] - Test different recovery criteria (_Florian?_)
- [ ] - Inspection of LCs (_Catriona + Florian_)
- [ ] - Stats! - chat to Marko Sestovic? (_Catriona + Florian_)
- [ ] - Flare detection map (flare amplitude)? (_Catriona / Florian_)

**Longer timeline: (Nice to have)**
- [ ] - Maybe compare flare detection methods
- [ ] - Rotational periods statistics
- [ ] - Maybe compare using GP with rotational kernel for detrending

**Comment by catrionamurray** [comment]
Speed of injection/recovery appears to be ~linear with number of injected planets - each planet injection and recovery takes ~5.5s for 1200 data points
<img width="575" alt="Screenshot 2024-07-10 at 5 40 51 PM" src="https://github.com/catrionamurray/occultence/assets/34891147/e2614ebc-3d16-4a45-af4e-0c782300a7b9">

**Reply by catrionamurray** [reply]
Time scales with number of data points (modeled with n=200 segments separated by fake night gaps) ~linearly. The gradient is 1.2 s/ 1000 data points:
![Screenshot 2024-07-12 at 3 03 39 PM](https://github.com/user-attachments/assets/55dfa3bd-dc0c-4cc1-80f5-31e93dacd55b)

**Reply by catrionamurray** [reply]
Out of the ~5s it takes to do one inj-rec BLS takes ~2s and we perform it twice = 4s. Once pre-detrending to avoid biasing the detrending and once to search for planets post-detrending.

**Reply by catrionamurray** [reply]
I tried added the option for multiprocessing/pooling for the recovery, however, pooling the entire process is currently slower than running recovery sequentially. I believe there's a bug in the pooling of the GP detrending (_might be related to NaNs? Or might be related to [this issue](https://github.com/dfm/george/issues/133)_). If I just pool the 2 BLS searches we save a lot of time.

**Reply by catrionamurray** [reply]
Demonstration of with/without pooling (assuming 5.5s per inj-rec planet without pooling) - it's around 3x as fast now on my 10 cores:
<img width="601" alt="Screenshot 2024-07-16 at 12 49 52 PM" src="https://github.com/user-attachments/assets/859cdce3-3210-451f-9bc5-1c1ae880f7c6">

**Reply by catrionamurray** [reply]
for ~1000 injection-recovery it's currently taking 30 minutes so for 180 targets with 2000 injection-recoveries it should be 180 hours = 7.5 days (for 1000 data points). Presumably @florian-lienhard  light curves are larger so 90d should now be <30d.

**Reply by florian-lienhard** [reply]
I assume 90d refers to the running time estimate we had previously. Alright, this sounds more manageable...

**Comment by catrionamurray** [comment]
The develop branch (https://github.com/catrionamurray/occultence/tree/develop) has the `pool_bls` option working

**Comment by catrionamurray** [comment]
In a new offshoot branch (https://github.com/catrionamurray/occultence/tree/develop_flare_detection) I have added flare injection and Lomb-scargle rotation period detection and detrending. Still working on adding flare detection.

**Reply by catrionamurray** [reply]
Added a very simple sigma-clipping detection method in this branch:
<img width="1010" alt="Screenshot 2024-07-18 at 3 04 35 PM" src="https://github.com/user-attachments/assets/9a0b2403-e114-4126-837b-114a284cc5d9">

**Reply by catrionamurray** [reply]
An example of the LS detrending and flare detection/masking is shown in this notebook: https://github.com/catrionamurray/occultence/blob/develop_flare_detection/occultence/notebooks/rotation_and_flares.ipynb
